pkg/store/cache/caching_bucket.go

```diff
@@ -0,0 +1,444 @@
+		ChunkObjectSizeTTL:        24 * time.Hour,
+		ChunkBlockTTL:             24 * time.Hour,
+		MaxChunksGetRangeRequests: 3,
+	}
+}
+
+// Bucket implementation that provides some caching features, using knowledge about how Thanos accesses data.
+type CachingBucket struct {
+	bucket objstore.Bucket
+	cache  cache.Cache
+
+	config CachingBucketConfig
+
+	logger log.Logger
+
+	requestedChunkBytes prometheus.Counter
+	fetchedChunkBytes   *prometheus.CounterVec
+	refetchedChunkBytes *prometheus.CounterVec
+
+	objectSizeRequests prometheus.Counter
+	objectSizeHits     prometheus.Counter
+}
+
+func NewCachingBucket(b objstore.Bucket, c cache.Cache, chunks CachingBucketConfig, logger log.Logger, reg prometheus.Registerer) (*CachingBucket, error) {
+	if b == nil {
+		return nil, errors.New("bucket is nil")
+	}
+	if c == nil {
+		return nil, errors.New("cache is nil")
+	}
+
+	cb := &CachingBucket{
+		config: chunks,
+		bucket: b,
+		cache:  c,
+		logger: logger,
+
+		requestedChunkBytes: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_requested_chunk_bytes_total",
+			Help: "Total number of requested bytes for chunk data.",
+		}),
+		fetchedChunkBytes: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_fetched_chunk_bytes_total",
+			Help: "Total number of fetched chunk bytes. Data from bucket is then stored to cache.",
+		}, []string{"origin"}),
+		refetchedChunkBytes: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_refetched_chunk_bytes_total",
+			Help: "Total number of chunk bytes re-fetched from storage, despite being in cache already.",
+		}, []string{"origin"}),
+		objectSizeRequests: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_objectsize_requests_total",
+			Help: "Number of object size requests for objects.",
+		}),
+		objectSizeHits: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_store_bucket_cache_objectsize_hits_total",
+			Help: "Number of object size hits for objects.",
+		}),
+	}
+
+	cb.fetchedChunkBytes.WithLabelValues(originBucket)
+	cb.fetchedChunkBytes.WithLabelValues(originCache)
+	cb.refetchedChunkBytes.WithLabelValues(originCache)
+
+	return cb, nil
+}
+
+func (cb *CachingBucket) Close() error {
+	return cb.bucket.Close()
+}
+
+func (cb *CachingBucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	return cb.bucket.Upload(ctx, name, r)
+}
+
+func (cb *CachingBucket) Delete(ctx context.Context, name string) error {
+	return cb.bucket.Delete(ctx, name)
+}
+
+func (cb *CachingBucket) Name() string {
+	return "caching: " + cb.bucket.Name()
+}
+
+func (cb *CachingBucket) WithExpectedErrs(expectedFunc objstore.IsOpFailureExpectedFunc) objstore.Bucket {
+	if ib, ok := cb.bucket.(objstore.InstrumentedBucket); ok {
+		// Make a copy, but replace bucket with instrumented one.
+		res := &CachingBucket{}
+		*res = *cb
```

Comment:
Do we nead a copy?

Reply:
This code copies cache, config, logger and metrics, but replaces bucket with "instrumented one". Without doing copy, we would need to copy all fields manually.